sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleHistogramAggregator.java

```diff
@@ -0,0 +1,174 @@
+      Resource resource,
+      InstrumentationLibraryInfo instrumentationLibraryInfo,
+      InstrumentDescriptor instrumentDescriptor,
+      double[] boundaries,
+      boolean stateful) {
+    super(resource, instrumentationLibraryInfo, instrumentDescriptor, stateful);
+    this.boundaries = boundaries;
+  }
+
+  @Override
+  public AggregatorHandle<HistogramAccumulation> createHandle() {
+    return new Handle(this.boundaries);
+  }
+
+  @Override
+  public final HistogramAccumulation merge(HistogramAccumulation x, HistogramAccumulation y) {
+    if (!x.getBoundaries().equals(y.getBoundaries())) {
+      throw new IllegalArgumentException("can't merge histograms with different boundaries");
+    }
+
+    long[] mergedCounts = new long[x.getCounts().size()];
+    for (int i = 0; i < x.getCounts().size(); ++i) {
+      mergedCounts[i] = x.getCounts().get(i) + y.getCounts().get(i);
+    }
+    return HistogramAccumulation.create(
+        x.getCount() + y.getCount(),
+        x.getSum() + y.getSum(),
+        x.getBoundaries(),
+        Arrays.stream(mergedCounts).boxed().collect(Collectors.toList()));
+  }
+
+  @Override
+  public final MetricData toMetricData(
+      Map<Labels, HistogramAccumulation> accumulationByLabels,
+      long startEpochNanos,
+      long lastCollectionEpoch,
+      long epochNanos) {
+    return MetricData.createDoubleHistogram(
+        getResource(),
+        getInstrumentationLibraryInfo(),
+        getInstrumentDescriptor().getName(),
+        getInstrumentDescriptor().getDescription(),
+        getInstrumentDescriptor().getUnit(),
+        DoubleHistogramData.create(
+            isStateful() ? AggregationTemporality.CUMULATIVE : AggregationTemporality.DELTA,
+            MetricDataUtils.toDoubleHistogramPointList(
+                accumulationByLabels,
+                isStateful() ? startEpochNanos : lastCollectionEpoch,
+                epochNanos)));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateDouble(double value) {
+    return HistogramAccumulation.create(
+        1, value, Collections.emptyList(), Collections.singletonList(1L));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateLong(long value) {
+    return HistogramAccumulation.create(
+        1, value, Collections.emptyList(), Collections.singletonList(1L));
+  }
+
+  static final class Handle extends AggregatorHandle<HistogramAccumulation> {
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    @GuardedBy("lock")
+    private final State current;
+
+    Handle(double[] boundaries) {
+      current = new State(boundaries);
+    }
+
+    @Override
+    protected HistogramAccumulation doAccumulateThenReset() {
+      lock.writeLock().lock();
+      try {
+        long[] cumulative = new long[current.counts.length];
+        cumulative[0] = current.counts[0];
+        for (int i = 1; i < current.counts.length; ++i) {
+          cumulative[i] = cumulative[i - 1] + current.counts[i];
```

Comment:
I don't think this is right for OTel histograms (at least not for OTLP). I see in the proto definition: 

`The sum of the bucket_counts must equal the value in the count field.`

If I read this code correctly, you've implemented this like the prometheus histogram, where each bucket count includes the count from the bucket below. We should keep the implementation here to match the OTLP as closely as possible, and then map to the prometheus format in that exporter.

Reply:
Thanks for pointing it out. I read that description but it slipped away from my mind while coding 😅 .